src/tests/suite/daos_obj.c

```diff
@@ -3365,6 +3365,171 @@ io_pool_map_refresh_trigger(void **state)
 	ioreq_fini(&req);
 }
 
+/**
+ * Test to fetch the non existence keys.
+ */
+static void nonexistent_internal(void **state, daos_obj_id_t oid,
+				 const char key_type[])
+{
+	test_arg_t	*arg = *state;
+	struct ioreq	req;
+	char		*update_buf;
+	char		*fetch_buf;
+	unsigned int	size   = IO_SIZE_NVME;
+
+	D_ALLOC(update_buf, size);
+	assert_non_null(update_buf);
+	dts_buf_render(update_buf, size);
+	ioreq_init(&req, arg->coh, oid, DAOS_IOD_ARRAY, arg);
+	D_ALLOC(fetch_buf, size);
+	assert_non_null(fetch_buf);
+	memset(fetch_buf, 0, size);
+
+	/* Insert single dkey/akey,skip insert record for oid*/
+	if (strcmp(key_type, "oid") != 0) {
+		insert_single("exist dkey", "exist akey", 0, update_buf, size,
+			DAOS_TX_NONE, &req);
+	}
+
+	if (strcmp(key_type, "akey") == 0) {
+		/*Fetch non existent akey*/
+		lookup_single("exist dkey", "nonexistent akey", 0, fetch_buf,
+			      size, DAOS_TX_NONE, &req);
+	} else if (strcmp(key_type, "dkey") == 0) {
+		/*Fetch non existent dkey*/
+		lookup_single("nonexistent dkey", "exist akey", 0, fetch_buf,
+			      size, DAOS_TX_NONE, &req);
+	} else if (strcmp(key_type, "oid") == 0) {
+		/*Fetch non existent oid*/
+		lookup_single("nonexistent dkey", "nonexistent akey", 0,
+			      fetch_buf, size, DAOS_TX_NONE, &req);
+	}
+
+	/**
+	* As per current implementation non existing keys
+	* will not use the buffer at all during fetch
+	*/
+	assert_false(fetch_buf[0] != '\0');
+
+	D_FREE(update_buf);
+	D_FREE(fetch_buf);
+
+	ioreq_fini(&req);
+}
+
+/**
+ * Fetch mixed existing and nonexistent akey with variable
+ * record sizes for both NVMe and SCM.
+ */
+static void fetch_mixed_keys_internal(void **state, daos_obj_id_t oid,
+				      unsigned int    size,
+				      daos_iod_type_t iod_type,
+				      const char dkey[], const char akey[])
+{
+	test_arg_t   *arg = *state;
+	struct ioreq req;
+	char         *akeys[MANYREC_NUMRECS];
+	char         *rec[MANYREC_NUMRECS];
+	daos_size_t  rec_size[MANYREC_NUMRECS];
+	int          rx_nr[MANYREC_NUMRECS];
+	daos_off_t   offset[MANYREC_NUMRECS];
+	char         *val[MANYREC_NUMRECS];
+	daos_size_t  val_size[MANYREC_NUMRECS];
+	int          i;
+
+	ioreq_init(&req, arg->coh, oid, iod_type, arg);
+
+	for (i = 0; i < MANYREC_NUMRECS; i++) {
+		akeys[i] = calloc(30, 1);
+		assert_non_null(akeys[i]);
+		snprintf(akeys[i], 30, "%s%d", akey, i);
+		D_ALLOC(rec[i], size);
+		assert_non_null(rec[i]);
+		dts_buf_render(rec[i], size);
+		rec_size[i] = size;
+		rx_nr[i]    = 1;
+		offset[i]   = i * size;
+		val[i]      = calloc(size, 1);
+		assert_non_null(val[i]);
+		val_size[i] = size;
+	}
+
+	/** Insert */
+	insert(dkey, MANYREC_NUMRECS, (const char **)akeys, rec_size, rx_nr,
+	       offset, (void **)rec, DAOS_TX_NONE, &req);
+
+	/* update the non existent akeys*/
+	snprintf(akeys[1], 30, "%s8", akey);
```

Comment:
if somebody increases MANYREC_NUMRECS to more than 8, this will fail.
So it's better to completely change the format of the non_existent akeys.

Reply:
I will change the format